File: packages/types/src/seo.ts
import { z } from 'zod';

// ─── Request ──────────────────────────────────────────────────────────────────

export const seoStrategyRequestSchema = z.object({
  topic: z
    .string()
    .trim()
    .min(5, 'Please describe your business or service (at least 5 characters)')
    .max(400, 'Description must be 400 characters or fewer'),
  audience: z.string().max(200).default(''),
  market: z.string().max(80).default('English'),
  goal: z.enum(['traffic', 'leads', 'content', 'visibility']).default('leads'),
  url: z.string().max(200).default(''),
  useWebContext: z.boolean().default(false),
  _honey: z.string().default(''),
}).strict();

export type SeoStrategyRequest = z.infer<typeof seoStrategyRequestSchema>;

// ─── Response ─────────────────────────────────────────────────────────────────
// Mirrors apps/ai/schemas/seo.py's _LenientModel classes: response_format=
// json_object only guarantees valid JSON syntax, not schema conformance, so
// the backend deliberately keeps most fields as plain strings with safe
// defaults (extra="allow") rather than strict enums — a live model phrasing
// e.g. intent_fit slightly differently than the prompt's prose must still
// validate, not fail closed. This schema mirrors that leniency exactly
// (z.string() where the backend has str, .passthrough() not .strict()) —
// tightening it here would silently reject live output the backend itself
// accepts, which is worse than not validating at all. rank/opportunity_score
// keep their numeric bounds because those, unlike the string fields, are an
// exact contract (_STEP2_SCHEMA) with no phrasing-drift risk.

export type SeoRelevance = 'high' | 'medium' | 'low';
export type SeoIntentFit = 'strong' | 'moderate' | 'weak';
export type SeoContentFormat =
  | 'landing page'
  | 'guide'
  | 'comparison article'
  | 'FAQ'
  | 'case study'
  | 'checklist'
  | 'tutorial';

const businessContextSchema = z.object({
  business_type: z.string().default(''),
  core_service: z.string().default(''),
  target_audience: z.string().default(''),
  value_proposition: z.string().default(''),
  pain_points: z.array(z.string()).default([]),
  competitive_differentiators: z.array(z.string()).default([]),
}).passthrough();

export type BusinessContext = z.infer<typeof businessContextSchema>;

const keywordCandidateSchema = z.object({
  term: z.string().default(''),
  type: z.string().default(''),
  intent: z.string().default(''),
}).passthrough();

export type KeywordCandidate = z.infer<typeof keywordCandidateSchema>;

const intentClusterSchema = z.object({
  intent: z.string().default(''),
  description: z.string().default(''),
  terms: z.array(z.string()).default([]),
  business_relevance: z.string().default(''),
}).passthrough();

export type IntentCluster = z.infer<typeof intentClusterSchema>;

const rankedOpportunitySchema = z.object({
  rank: z.number().int().min(1).max(10).default(1),
  term: z.string().default(''),
  intent: z.string().default(''),
  opportunity_score: z.number().int().min(0).max(100).default(0),
  lead_relevance: z.string().default(''),
  business_relevance: z.string().default(''),
  intent_fit: z.string().default(''),
  content_gap_potential: z.string().default(''),
  conversion_closeness: z.string().default(''),
  suggested_content_format: z.string().default(''),
  cta_angle: z.string().default(''),
  why_ranked_here: z.string().default(''),
}).passthrough();

export type RankedOpportunity = z.infer<typeof rankedOpportunitySchema>;

const contentIdeaSchema = z.object({
  title: z.string().default(''),
  format: z.string().default(''),
  target_terms: z.array(z.string()).default([]),
  rationale: z.string().default(''),
}).passthrough();

export type ContentIdea = z.infer<typeof contentIdeaSchema>;

const leadAngleSchema = z.object({
  angle: z.string().default(''),
  cta: z.string().default(''),
  target_terms: z.array(z.string()).default([]),
  rationale: z.string().default(''),
}).passthrough();

export type LeadAngle = z.infer<typeof leadAngleSchema>;

const roadmapPhaseSchema = z.object({
  phase: z.string().default(''),
  focus: z.string().default(''),
  items: z.array(z.string()).default([]),
  rationale: z.string().default(''),
}).passthrough();

export type RoadmapPhase = z.infer<typeof roadmapPhaseSchema>;

export const seoStrategyResultSchema = z.object({
  mode: z.enum(['live', 'mock']),
  summary: z.string().default(''),
  extracted_business_context: businessContextSchema.default({}),
  keyword_candidates: z.array(keywordCandidateSchema).default([]),
  intent_clusters: z.array(intentClusterSchema).default([]),
  reranked_opportunities: z.array(rankedOpportunitySchema).default([]),
  content_ideas: z.array(contentIdeaSchema).default([]),
  lead_generation_angles: z.array(leadAngleSchema).default([]),
  roadmap: z.array(roadmapPhaseSchema).default([]),
  warnings: z.array(z.string()).default([]),
  metadata: z.object({
    model: z.string(),
    used_web_context: z.boolean(),
    market: z.string(),
    goal: z.enum(['traffic', 'leads', 'content', 'visibility']),
    disclaimer: z.string(),
  }).passthrough(),
}).passthrough();

export type SeoStrategyResult = z.infer<typeof seoStrategyResultSchema>;

File: apps/ai/schemas/seo.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class SeoStrategyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    topic: str
    audience: str = ""
    market: str = "English"
    goal: Literal["traffic", "leads", "content", "visibility"] = "leads"
    url: str = ""

    @field_validator("topic")
    @classmethod
    def validate_topic(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 5:
            raise ValueError("Topic must be at least 5 characters")
        if len(v) > 400:
            raise ValueError("Topic must be 400 characters or fewer")
        return v

    @field_validator("audience")
    @classmethod
    def validate_audience(cls, v: str) -> str:
        return v.strip()[:200]

    @field_validator("market")
    @classmethod
    def validate_market(cls, v: str) -> str:
        return v.strip()[:80]

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: str) -> str:
        v = v.strip()
        if not v:
            return v
        if len(v) > 200:
            raise ValueError("URL must be 200 characters or fewer")
        return v


# ─── LLM output models (Step 1 / Step 2) ───────────────────────────────────
# These calls use response_format={"type": "json_object"} (valid-JSON-syntax
# guarantee only — no schema enforcement, unlike CS01/CS02's strict:true
# tool-calling), so unlike those, Pydantic here is the *only* structural
# check, not a second one on top of an API-level guarantee. Deliberately
# lenient — extra="allow", every field optional with a safe default, plain
# str/int instead of Literal enums — so a live response that's close to but
# not byte-for-byte the documented shape (_STEP1_SCHEMA / _STEP2_SCHEMA in
# routes/seo.py) still validates. The goal is catching genuinely malformed
# output (wrong top-level shape, wrong types) without becoming a new source
# of live failures over a model phrasing an enum-ish field slightly
# differently than the prompt's prose description asked for.


class _LenientModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class ExtractedBusinessContext(_LenientModel):
    business_type: str = ""
    core_service: str = ""
    target_audience: str = ""
    value_proposition: str = ""
    pain_points: list[str] = Field(default_factory=list)
    competitive_differentiators: list[str] = Field(default_factory=list)


class KeywordCandidate(_LenientModel):
    term: str = ""
    type: str = ""
    intent: str = ""


class IntentCluster(_LenientModel):
    intent: str = ""
    description: str = ""
    terms: list[str] = Field(default_factory=list)
    business_relevance: str = ""


class Step1Output(_LenientModel):
    extracted_business_context: ExtractedBusinessContext = Field(
        default_factory=ExtractedBusinessContext
    )
    keyword_candidates: list[KeywordCandidate] = Field(default_factory=list)
    intent_clusters: list[IntentCluster] = Field(default_factory=list)
    summary: str = ""


class RerankedOpportunity(_LenientModel):
    rank: int = 0
    term: str = ""
    intent: str = ""
    opportunity_score: int = 0
    lead_relevance: str = ""
    business_relevance: str = ""
    intent_fit: str = ""
    content_gap_potential: str = ""
    conversion_closeness: str = ""
    suggested_content_format: str = ""
    cta_angle: str = ""
    why_ranked_here: str = ""


class ContentIdea(_LenientModel):
    title: str = ""
    format: str = ""
    target_terms: list[str] = Field(default_factory=list)
    rationale: str = ""


class LeadGenerationAngle(_LenientModel):
    angle: str = ""
    cta: str = ""
    target_terms: list[str] = Field(default_factory=list)
    rationale: str = ""


class RoadmapPhase(_LenientModel):
    phase: str = ""
    focus: str = ""
    items: list[str] = Field(default_factory=list)
    rationale: str = ""


class Step2Output(_LenientModel):
    reranked_opportunities: list[RerankedOpportunity] = Field(default_factory=list)
    content_ideas: list[ContentIdea] = Field(default_factory=list)
    lead_generation_angles: list[LeadGenerationAngle] = Field(default_factory=list)
    roadmap: list[RoadmapPhase] = Field(default_factory=list)


class CorrectedSeoOutput(Step1Output, Step2Output):
    """_correction_pass is asked to return the complete result — every
    Step1Output and Step2Output field together — with issues fixed in
    place, not just one step's slice of it. Previously that response was
    only checked with a bare `corrected.get("reranked_opportunities")`
    truthiness test, unlike the original step1/step2 calls this class
    otherwise mirrors. extra="allow" (inherited) preserves mode/warnings/
    metadata, which the correction prompt also echoes back but which
    aren't part of either step's own output shape.
    """
